### PR TITLE
Initial Search Application Search API with templates

### PR DESCRIPTION
**Pull request by carlosdelest**
Adds a search endpoint to Search Applications:

`GET /_application/search_application/<name>/_search`
```json
{
    "params": {
        "param1": "value1",
        "param2": "value2"
    }
}
```

where `params` contains overrides for the Search Application defined template, for example:

`PUT /_application/search_application/<name>`
```json
{
    "indices": ["test1"],
    "template": {
        "script": {
            "source": {
                "query": {
                    "term": {
                        "{{param1}}": "{{param2}}"
                    }
                }
            },
            "params": {
                "param1": "hello",
                "param2": "world"
            }
        }            
    }
}
```

For the above Search Application and search request, the template would render like:

```json
              "query": {
                    "term": {
                        "value1": "value2"
                    }
                }
```

... as the search specified parameters would override the template ones.

Based on [this work](https://github.com/elastic/elasticsearch/pull/94498) by Jim and Ioana 💯

### CODE REVIEW

File: x-pack/plugin/ent-search/qa/rest/roles.yml [hunk 1]
```diff
@@ -22,5 +22,11 @@ entsearch:
         "test-re-creating-search-application",
         "test-search-application-to-delete",
         "test-nonexistent-search-application",
-      ]
+    ]
       privileges: [ "manage" ]
+    - names: [
+      # indices used for indexing and searching
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/qa/rest/roles.yml:28]: Some indices for the search use case, with read and write permissions

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java [hunk 2]
```diff
@@ -42,48 +42,26 @@ public class SearchApplication implements Writeable, ToXContentObject {
...
-     * @param analyticsCollectionName The name of the associated analytics collection.
-     * @param updatedAtMillis         Last updated time in milliseconds for the search application.
+     * @param name                      The name of the search application.
+     * @param indices                   The list of indices targeted by this search application.
+     * @param analyticsCollectionName   The name of the associated analytics collection.
+     * @param updatedAtMillis           Last updated time in milliseconds for the search application.
+     * @param searchApplicationTemplate The search application template to be used on search
      */
-    public SearchApplication(String name, String[] indices, @Nullable String analyticsCollectionName, long updatedAtMillis) {
+    public SearchApplication(
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java:58]: A little messy rebase on the `ent-search-module` branch 😱 . Added the Search Template to the application.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class SearchApplicationQueryParams implements ToXContentObject, Writeable {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java:25]: Created a new class for the query parameters. Just now it's simple enough, but we could potentially extend it with other options / params in the future, so I decided to create a separate class for handling this

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/RestQuerySearchApplicationAction.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, ENDPOINT_PATH), new Route(POST, ENDPOINT_PATH));
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
+        SearchApplicationQueryParams queryParams = new SearchApplicationQueryParams();
+        if (restRequest.hasContent()) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/RestQuerySearchApplicationAction.java:41]: This leaves open the possibility of not including a body at all, and just issue the request with no body.

File: rest-api-spec/src/main/resources/rest-api-spec/api/search_application.query.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
+{
+  "search_application.query": {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/search_application.query.json:2]: Good catch!

Right now, it matches the action names (`RestQuerySearchApplicationAction`, `QuerySearchApplicationAction`, `TransportQuerySearchApplicationAction`). I guess `Query` is a better name as `SearchSearchApplicationAction` would be very confusing.

I'm in favour of refactoring these names and drop the SearchApplication from them, but that's something that can wait for later.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java (hunk 3 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java:29]: Can it be implemented in the `Request` directly? The other params you're talking about are would be for the request so I guess we can add everything there?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private final Map<String, Object> templateParams;
+
+    public SearchApplicationQueryParams() {
+        this(Map.of());
+    }
+
+    public SearchApplicationQueryParams(Map<String, Object> templateParams) {
+        // TODO Understand how can we make the parser build the object using the default constructor instead of this one
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java:52]: Why is it needed?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+/**
+ * Search template included in a {@link SearchApplication}. It will be used for searching using the
+ * {@link org.elasticsearch.xpack.application.search.action.QuerySearchApplicationAction}, overriding the parameters included on it via
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:26]: Let's import this to get rid of the package name?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ */
+public class SearchApplicationTemplate implements ToXContentObject, Writeable {
+    private final Script script;
+
+    public SearchApplicationTemplate(StreamInput in) throws IOException {
+        this.script = in.readOptionalWriteable(Script::new);
+    }
+
+    public SearchApplicationTemplate(Script script) {
+        this.script = script;
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:37]: Should we validate that the script language is correctly set to `mustache`?

**Review by jimczi — APPROVED** [review]
Looks great @carlosdelest!  I left some minor comments.

File: rest-api-spec/src/main/resources/rest-api-spec/api/search_application.query.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/search_application.query.json]: Changed to search_application.search, thanks for the discussion @sethmlarson !

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java (hunk 7 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java]: Done, thanks!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java (hunk 6 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java]: I changed the parsing code - this is no longer an issue.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java (hunk 3 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationQueryParams.java]: I think you're right. Changed it to be included in the Request.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java (hunk 8 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:38]: Added 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/GetSearchApplicationAction.java [hunk 9]
```diff
@@ -91,8 +92,14 @@ public Response(SearchApplication app) {
...
         }
 
-        public Response(String name, String[] indices, String analyticsCollectionName, long updatedAtMillis) {
-            this.searchApp = new SearchApplication(name, indices, analyticsCollectionName, updatedAtMillis);
+        public Response(
+            String name,
+            String[] indices,
+            String analyticsCollectionName,
+            long updatedAtMillis,
+            SearchApplicationTemplate searchApplicationTemplate
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/GetSearchApplicationAction.java:100]:
```suggestion
            SearchApplicationTemplate template
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/GetSearchApplicationAction.java [hunk 10]
```diff
@@ -91,8 +92,14 @@ public Response(SearchApplication app) {
...
+        ) {
+            this.searchApp = new SearchApplication(name, indices, analyticsCollectionName, updatedAtMillis, searchApplicationTemplate);
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/GetSearchApplicationAction.java:102]:
```suggestion
            this.searchApp = new SearchApplication(name, indices, analyticsCollectionName, updatedAtMillis, template);
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/QuerySearchApplicationAction.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+        public Request(String name) {
+            this(name, Map.of());
+        }
+
+        public Request(String name, Map<String, Object> queryParams) {
+            Objects.requireNonNull(name, "Application name must be specified");
+            this.name = name;
+
+            Objects.requireNonNull(name, "Query parameters must be specified");
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/QuerySearchApplicationAction.java:72]:
```suggestion
            Objects.requireNonNull(queryParams, "Query parameters must be specified");
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/TransportQuerySearchApplicationAction.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TransportQuerySearchApplicationAction extends SearchApplicationTransportAction<
+    QuerySearchApplicationAction.Request,
+    SearchResponse> {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/TransportQuerySearchApplicationAction.java:41]: I think it's OK to return a `SearchResponse` in this PR, but some day in the future we may want to tack on additional data to a custom response that includes a search response.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/TransportQuerySearchApplicationAction.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                final SearchSourceBuilder sourceBuilder = renderTemplate(script, mergeTemplateParams(request, script));
+                SearchRequest searchRequest = new SearchRequest(searchApplication.indices()).source(sourceBuilder);
+
+                client.execute(
+                    SearchAction.INSTANCE,
+                    searchRequest,
+                    listener.delegateFailure((l2, searchResponse) -> l2.onResponse(searchResponse))
+                );
+            } catch (IOException exc) {
+                l.onFailure(exc);
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/action/TransportQuerySearchApplicationAction.java:93]:
```suggestion
            } catch (IOException e) {
                l.onFailure(e);
```

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/search/SearchApplicationTestUtils.java [hunk 14]
```diff
@@ -34,7 +44,25 @@ public static SearchApplication randomSearchApplication() {
             ESTestCase.randomAlphaOfLengthBetween(1, 10),
             generateRandomStringArray(10, 10, false, false),
             randomFrom(new String[] { null, randomAlphaOfLengthBetween(1, 10) }),
-            randomLongBetween(0, Long.MAX_VALUE)
+            randomLongBetween(0, Long.MAX_VALUE),
+            randomBoolean() ? getRandomSearchApplicationTemplate() : null
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/search/SearchApplicationTestUtils.java:48]: Nice 👍

**Review by kderusso — APPROVED** [review]
Looks great! One change that needs to happen in `QuerySearchApplicationAction`, everything else is non-blocking nitpicky comments. Thank you!

**Pull request closed without merging**